### DISCUSSION TITLE
Setting up Go to definition?

### DISCUSSION THREAD

**Discussion by MarcCoquand** [mrcjkb/haskell-tools.nvim#205] [Q&A] [answered]
Hi!

I can't really understand how I am supposed to use the go to definition functionality. Using `gd` does not seem to work. The documentation only mentions that it's possible, but not how to set it up.

**Comment by mrcjkb** [comment, marked as answer]
Hey,

if you want to map `gd` to go-to-definition, you can add a keymap:

```lua
opts = { noremap = true, silent = true, buffer = bufnr } -- bufnr is an argument of the on_attach function
vim.keymap.set('n', 'gd', vim.lsp.buf.definition, opts)
```

I don't include all LSP-related keymaps in the readme, because there are [already a bunch of suggestions in the nvim-lspconfig readme](https://github.com/neovim/nvim-lspconfig#suggested-configuration) (which I link to).

For go-to-definition specifically, neovim automatically sets up a [`tagfunc`](https://neovim.io/doc/user/lsp.html#vim.lsp.tagfunc()), so you can use [`C-[`](https://neovim.io/doc/user/tagsrch.html#CTRL-%5D) to jump to a definition, and fall back to tags if the definition can't be found. This is useful if you also use something like [fast-tags](https://hackage.haskell.org/package/fast-tags).

**Reply by MarcCoquand** [reply]
Ah, I have that set up, but I didn't need to explicitly set it for other languages. It seems to work now when I added it to my ftplugin/haskell file.

**Reply by mrcjkb** [reply]
Great to hear you were able to solve it :smile: 

I guess for the other languages, you had it set in the `on_attach` function?

**Reply by MarcCoquand** [reply]
Yeah exactly, and it seems to haskell-tools has a separate procedure for how it attaches?

Thank you so much for the support :)

**Reply by mrcjkb** [reply]
Both haskell-tools and nvim-lspconfig use [`vim.lsp.start()`](https://neovim.io/doc/user/lsp.html) under the hood, so it's the same procedure.
But nvim-lspconfig sets up autocommands for filetypes, in order to detect when to attach which LSP client.
haskell-tools can do that, too (with the `setup()` function), but I recommend to use `setup_or_attach()` in a ftplugin/haskell.lua file, because setting up autocommands slows down Neovim's startup performance.

haskell-tools won't use anything from nvim-lspconfig, unless you explicitly pass it into the `on_attach` callback.
